system/repositories.rs

```diff
@@ -0,0 +1,143 @@
+        },
+    };
+
+    pub fn list(tx: TransactionRead<WALClient>) -> Vec<User> {
+        let query = parse_query("match (user: $u, password: $p) isa user-password; $u has name $n;").unwrap();
+        let (tx, result) = execute_read_pipeline(tx, &query.into_pipeline());
+        let rows = result.unwrap();
+        let users = rows.iter().map(|row| User::new(get_string(&tx, &row, "n"))).collect();
+        users
+    }
+
+    pub fn get(tx: TransactionRead<WALClient>, username: &str) -> Option<(User, Credential)> {
+        let query = parse_query(
+            format!(
+                "match
+                (user: $u, password: $p) isa user-password;
+                $u has name '{username}';
+                $p has hash $h;"
+            )
+            .as_str(),
+        )
+        .unwrap();
+        let (tx, result) = execute_read_pipeline(tx, &query.into_pipeline());
+        let mut rows: Vec<HashMap<String, VariableValue>> = result.unwrap();
+        if !rows.is_empty() {
+            let row = rows.pop().unwrap();
+            let hash = get_string(&tx, &row, "h");
+            Some((User::new(username.to_string()), Credential::PasswordType { password_hash: PasswordHash::new(hash) }))
+        } else {
+            None
+        }
+    }
+
+    pub fn create(
+        snapshot: WriteSnapshot<WALClient>,
+        type_manager: &TypeManager,
+        thing_manager: Arc<ThingManager>,
+        function_manager: &FunctionManager,
+        user: &User,
+        credential: &Credential,
+    ) -> Arc<WriteSnapshot<WALClient>> {
+        let query = match credential {
+            Credential::PasswordType { password_hash: PasswordHash { value: hash } } => {
+                let uuid = Uuid::new_v4().to_string();
+                parse_query(
+                    format!(
+                        "insert $u isa user, has uuid '{uuid}', has name '{name}';
+                        $p isa password, has hash '{hash}';
+                        (user: $u, password: $p) isa user-password;",
+                        uuid = uuid,
+                        name = user.name,
+                        hash = hash
+                    )
+                    .as_str(),
+                )
+                .unwrap()
+            }
+        };
+        let (_, snapshot) =
+            execute_write_pipeline(snapshot, type_manager, thing_manager, function_manager, &query.into_pipeline());
+        snapshot
+    }
+
+    pub fn update(
+        snapshot: WriteSnapshot<WALClient>,
+        type_manager: &TypeManager,
+        thing_manager: Arc<ThingManager>,
+        function_manager: &FunctionManager,
+        username: &str,
+        user: &Option<User>,
+        credential: &Option<Credential>,
+    ) -> Arc<WriteSnapshot<WALClient>> {
+        if user.is_some() {
+            todo!("update user detail")
+        }
+
+        if credential.is_some() {
+            todo!("update credential detail")
+        }
+
+        todo!()
+    }
+
+    pub fn delete(
```

Comment:
What do we expect when we delete a user which is logged in? Change their password? Do we need to disconnect them with an error directly? Do we expect all the API calls to be verified with an auth error if something changed between the connection and this call?

Reply:
When a user is deleted, subsequent connections made by them will be rejected. However, an already running transaction will stay open until it is closed. Given that transaction has a timeout, the user will not be able to maintain it indefinitely. I think this solution should be good enough for now, right @flyingsilverfin ?

Reply:
What about user's operations? Can they query, commit? All the permissions should be updated immediately. Otherwise we are not really ready production user management

Reply:
I agree with this, I think connections associated with a user that is either deleted, or has their credential updated should immediately be closed. However, in order to do that we need to keep track of which connections belong to which user, which is something the server  currently doesn't do.

Therefore, as have been [discussed](https://discord.com/channels/665254494820368395/672580553677078548/1311046300401401896) with @flyingsilverfin over Discord, we agreed that we'll postpone this issue as performance optimisation takes higher precedence.